### PR TITLE
Add `-FromUnixTime` to `Get-Date` to allow Unix time input

### LINKED ISSUE #11719: Add One New DateTime Parsing Format
Yesterday I posted a suggestion that we have a new special format for Get-Date that creates the SecondsSinceEpoch (aka UnixTime) for users.

Additionally I think we should have built into the language the ability to easily parse a UnixEpoch timestamp back into a DateTime.

Various existing parsing that works today:
```
$aDate = [DateTime]"2020-01-28T17:32:18.4986003-06:00" #  (conversion back from the format of Format "o")
$bDate = [DateTime]"2020-01-28T17:36:39" #  (conversion back from the format of Format "s")
$cDate = [DateTime]"Tue, 28 Jan 2020 17:37:43 GMT" #  (conversion back from the format of Format "r")
$dDate = [DateTime]"2020-01-28 17:38:24Z" #  (conversion back from the format of Format "u")
$eDate = [DateTime]"7/25/2013 6:37:31 PM" 
$fDate = [DateTime]"2013-07-25 14:26:00" 
$gDate = [DateTime]"7/25/2013 14:26:00"
$hDate = [DateTime]"2013-07-25 6:37:31 PM"
```
But we can't currently put in $AUnixDate = [DateTime]"1374755180" and end up with a valid DateTime object (which would be of UTC kind)

A discussion by Randy James at https://stackoverflow.com/questions/17859421/parse-datetime-in-multiple-formats suggests perhaps something along these lines for processing this in another language:
```
public DateTime ParseRequestDate()
{
    // https://stackoverflow.com/questions/2883576/how-do-you-convert-epoch-time-in-c

    CultureInfo enUS = new CultureInfo("en-US");

    var dt = "1374755180";
    //var dt = "7/25/2013 6:37:31 PM";
    //var dt = "2013-07-25 14:26:00";

    DateTime dateValue;
    long dtLong;

    // Scenario #1
    if (long.TryParse(dt, out dtLong))
        return dtLong.FromUnixTime();

    // Scenario #2 & #3
    var formatStrings = new string[] { "MM/dd/yyyy hh:mm:ss tt", "yyyy-MM-dd hh:mm:ss" };
    if (DateTime.TryParseExact(dt, formatStrings, enUS, DateTimeStyles.None, out dateValue))
        return dateValue;

    throw new SomeException("Don't know how to parse...");
}

```

### PR DESCRIPTION
**Pull request by jackdcasey**
# PR Summary

My proposal, is to add a  `-FromUnixTime` switch to Get-Date. 

This allows the input of a Unix timestamp into `Get-Date`. This is currently impossible, as the default unit for a numeric input is ticks. 

Adding this switch would allow the user to specify if the input is Unix time, while also keeping full previous compatibility with ticks. 

Here is an example output with, and without the switch: 
```
> Get-Date -Date 1584947314 -FromUnixTime

 Monday, March 23, 2020 7:08:34 AM

> Get-Date -Date 1584947314

 Monday, January 1, 0001 12:02:38 AM
```

## PR Context

Closes #11719

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [x] Issue filed: MicrosoftDocs/PowerShell-Docs#5727
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Date.Tests.ps1 [hunk 1]
```diff
@@ -277,3 +277,10 @@ Describe "Get-Date -UFormat tests" -Tags "CI" {
             Get-Date -Date $date -UFormat $format | Should -BeExactly $result
     }
 }
+
+Describe "Get-Date -FromUnixTime tests" -Tags "CI" {
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Date.Tests.ps1]: The `Describe` is no needed. You could put the test in first or second `Describe` in the file.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs [hunk 2]
```diff
@@ -547,6 +560,14 @@ private string UFormatDateString(DateTime dateTime)
             return StringUtil.Format(sb.ToString(), dateTime);
         }
 
+        /// <summary>
+        /// Converts unix time into DateTime.
+        /// </summary>
+        private static DateTime UnixTimeToDateTime(long unixTime)
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs]: The method is used only once and can be removed.

Comment (jackdcasey) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs]: Great point, updated! 😁

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Date.Tests.ps1 (hunk 1 above)
Comment (jackdcasey) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Date.Tests.ps1]: Makes sense, I've updated the tests!

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Date.Tests.ps1 [hunk 3]
```diff
@@ -172,6 +172,11 @@ Describe "Get-Date" -Tags "CI" {
         $timeDifference.Milliseconds | Should -BeLessThan 1
         $timeDifference.Ticks        | Should -BeLessThan 10000
     }
+
+    It "-FromUnixTime works" {
+        Get-Date -Date 1577836800 -FromUnixTime | Should -Be (Get-Date -Date 637134336000000000)
+        Get-Date -Date 0 -FromUnixTime | Should -Be (Get-Date -Date 621355968000000000)
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Date.Tests.ps1:178]: Please add a comment about these used constants.
Also right part should return UTC, not Local.

Comment (jackdcasey) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Date.Tests.ps1]: Good catch! I've added comments to those tests 😅

From what I can find, the `Get-Date -Date` cmdlet by itself will return a `DateTime` with the Kind set to `Unspecified`. Would adding the `-AsUTC` provide the functionality you're thinking of? 

For example, the tests would be: 
```
Get-Date -Date 1577836800 -FromUnixTime | Should -Be (Get-Date -Date 637134336000000000 -AsUTC)
Get-Date -Date 0 -FromUnixTime | Should -Be (Get-Date -Date 621355968000000000 -AsUTC)
```
Please let me know if this is what you're thinking!

Comment (iSazonov) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Date.Tests.ps1]: I'd expect AsUTC is right thing here.

Comment (jackdcasey) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Get-Date.Tests.ps1]: Makes sense. Updated!

**Comment by jackdcasey** [comment]
Upon specifying UTC in the tests, I noticed a small inconsistency. 

When specifying the DateTime with ticks, or with unix time, the `kind` was set as `Unspecified` instead of `Utc`. 

This was causing issues when combined with the `-AsUTC` parameter. When performing the conversion, the `ToUniversalTime()` method is called. This caused inconsistencies, as the local offset would be applied. The [documentation](https://docs.microsoft.com/en-us/dotnet/api/system.datetime.touniversaltime?view=net-5.0#remarks) for this method confirms this behaviour: 
```
Unspecified | The current DateTime object is assumed to be a local time, and the conversion is performed as if Kind were Local.
```
To get around this, we can set the `kind` to `Utc` explicitly when using ticks or unix time. I've made the following change to facilitate this: 

```
dateToUse = Date
```
to
```
dateToUse = DateTime.SpecifyKind(Date, DateTimeKind.Utc);
```

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs [hunk 4]
```diff
@@ -249,7 +249,7 @@ protected override void ProcessRecord()
                 }
                 else
                 {
-                    dateToUse = Date;
+                    dateToUse = DateTime.SpecifyKind(Date, DateTimeKind.Utc);
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs:252]: Why do we need the change? It looks like unrelated the PR.

Comment (jackdcasey) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs:252]: I added a few comments about this earlier, I wasn't sure if it'd be better to open a new PR for this change, or bundle it together with this one. 

This was done to address issues with using the `-AsUTC` parameter, with a `DateTime` with `kind` set as `Unspecified`. Due to the way `ToUniversalTime()` is handling the local offset, the resulting `DateTime` is off.  

Do you reckon this change should be added in a different PR?

Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs:252]: Yes, it is not related the PR.

Comment (jackdcasey) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs:252]: Sounds good, I'll revert this change and open a new PR 😁

Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs:252]: I think it will a breaking change. So please open new issue to discuss.

Comment (jackdcasey) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs]: Will do! I'll get that typed up now

Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs]: Sorry AsUtc was added in Preview1 and did not be jet released - no breaking change.

Comment (jackdcasey) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/GetDateCommand.cs]: I've submitted a new issue regarding this to get a conversation going, and ensure I'm interpreting things correctly 😁

**Review by iSazonov — APPROVED** [review]

**Comment by iSazonov** [comment]
@jackdcasey Please create new issue in PowerShell-Docs repo for the new parameter and add reference in the PR description. Thanks!

**Comment by jackdcasey** [comment]
Opened issue in MicrosoftDocs/PowerShell-Docs: MicrosoftDocs/PowerShell-Docs#5727

**Review by TravisEz13 — APPROVED** [review]

**Comment by iSazonov** [comment]
@jackdcasey Thanks for your contribution!

**Comment by jackdcasey** [comment]
@iSazonov Thank you for all the help!

**Comment by aetos382** [comment]
Negative UnixTime is not supported?

```
PS /> Get-Date -Date -1 -FromUnixTime
Get-Date: Cannot bind parameter 'Date'. Cannot convert value "-1" to type "System.DateTime". Error: "String '-1' was not recognized as a valid DateTime."
```

**Comment by iSazonov** [comment]
@aetos382 No. If you have a feature request please open new issue.
